src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+  const capData = marshaller.serialize(bridgeAction);
+  var data = JSON.stringify(capData)
+  return data
+};
+
+/**
+ * Function to check if currently in submission
+ * @param {*} feed feed to check for
+ * @returns whether last submission was made in less than SEND_CHECK_INTERVAL seconds
+ */
+const checkIfInSubmission = async (feed) => {
+
+  //get last submission time
+  let query = await queryTable("jobs", ["last_submission_time"], feed)
+  //get seconds since last price submission
+  let timePassedSinceSubmission = (Date.now()/1000) - 
+  query.last_submission_time
+  return timePassedSinceSubmission < Number(SEND_CHECK_INTERVAL)
+}
+
+/**
+  * Function to push price on chain to the smart wallet
+  * @param {*} price price to push
+  * @param {*} feed feed to push price to
+  * @param {*} round round to push result to
+  * @param {*} from account to push from
+  * @returns whether successful
+  */
+const pushPrice = async (price, feed, round, from) => {
+
+  //create an offerId with the Date number
+  var offerId = Date.now()
+
+  //get offers
+  let offers = await getOraclesInvitations()
+  //get previous offer for feed
+  let previousOffer = offers[feed]
+
+  //create an offer
+  const offer = {
+    id: Number(offerId),
+    invitationSpec: {
+      source: 'continuing',
+      previousOffer: Number(previousOffer),
+      invitationMakerName: 'PushPrice',
+      invitationArgs: harden([{ unitPrice: BigInt(price), roundId: round }]),
+    },
+    proposal: {},
+  };
+
+  //create keyring
+  var keyring = {
+    "home": "",
+    "backend": "test"
+  }
+
+  //check if submitted for round
+  let submitted = await checkSubmissionForRound(from, previousOffer, round)
+
+  //check if in submission
+  let inSubmission = await checkIfInSubmission(feed)
+
+  //loop retries
+  for (let i = 0; i < SUBMIT_RETRIES && !submitted && !inSubmission; i++) {
+
+    //query round
+    let latestRound = await queryRound(feed)
+    
+    //if latestRound is greater than round being pushed or submission to the round is already made, abort
+    if (latestRound.round_id > round || (latestRound.round_id == round && latestRound.submission_made)){
+      console.log("Price failed to be submitted for old round", round)
+      return false
+    }
+
+    console.log("Submitting price for round", round, "try", (i + 1))
+
+    offer.id = Number(Date.now())
+
+    //output action
+    var data = outputAction({
+      method: 'executeOffer',
+      // @ts-ignore
+      offer,
+    });
+
+    //change data to JSON
+    data = JSON.parse(data)
+
+    //execute
+    await execSwingsetTransaction(
+      "wallet-action --allow-spend '" + JSON.stringify(data) + "'",
+      networkConfig,
+      from,
+      false,
+      keyring,
+    );
+
+    //update last submission time
+    await updateTable("jobs", {"last_submission_time": Date.now()/1000}, "feed")
+
+    //sleep 13 seconds to wait 2 blocks and a bit
+    await delay((Number(SEND_CHECK_INTERVAL)+1) * 1000);
```

Comment:
Why wait for 2 blocks? Cosmos it is finalized after 1 block. Also the chain can be spammed and the tx is not yet reached by the miners. Is there a way from agoric sdk to know if a tx was mined? In ethers.js for example is:

```
const sendTransaction = async() => {
    const transaction = await provider.sendTransaction(rawTransaction);
    // wait() has the logic to return receipt once the transaction is mined
    const receipt = await wait(transaction);
}
```

Reply:
If agoric-sdk doesn't have, do a while loop retreiving the tx until it is mined or expired

Reply:
A transaction just being mined is not enough in this case, we want to make sure it is successful. The only way to know if it is successful is by querying the offers in the wallet as we are already doing. Regarding the 2 blocks, in the current testnet not even 2 blocks are enough for the price update to be confirmed. This may be due to the testnet being used as it is a local testnet with only 1 validator. We will monitor the time taken in devnet and mainnet and then adjust this parameter accordingly. This is not an issue as we just need to adjust this number. This can be up to 1 minute as it is fine waiting 1 minute to confirm the price update. We just do not want to send extra price submissions just because the price is not confirmed as this would result in extra costs for oracles.